Modular settings.

dhil opened the pull request:
Prior to the merge of effects and session #131 it was possible to
define settings on a per compilation unit basis, except for settings which
were settable via the commandline. They had to be defined early before
the CLI arguments were parsed.

This patch restores the ability to define settings on a compilation
unit basis, and it does so for all settings -- meaning commandline
settable settings can be defined locally too.

At glance this patch contains the following changes:

* Local definition of settings.
* Small DSL for defining settings and their CLI extensions.
* No universal type. Settings are now GADTs, meaning reading and
  writing operations on settings are typed. As a consequence values
  are no longer restricted by a universal type.
* GNU convention compatible handling of CLI arguments.
* Retrieval of commandline arguments in Links programs via `getArgs`.
* Automatic gathering and printing of CLI settings.

To illustrate the latter point, passing `-h` or `--help` to Links on
the commandline causes Links to print something along the lines of

```
usage: links.exe [options] [source-files [-- arguments]]

Options are:
     --config=<file>             Initialises Links according to the given configuration file
 -d, --debug                     Prints internal debugging information (development)
     --enable-handlers           Enables the effect handlers extension
 -e, --evaluate=<expression>     Evaluates an expression
 -h, --help                      Print help message and exit
     --path=<dir[[:dir']...]>    Search paths for Links modules
     --measure_performance       Instruments the server-side runtime to measure various performance characteristics
 -r, --rlwrap                    Toggles whether to use the native readline support in REPL mode
     --optimise                  Optimises the generated code
     --print-keywords            Print keywords and exit
     --session-exceptions        Enables the session exceptions extension
     --set=<setting=value>       Sets the value of a particular setting via the commandline
 -v, --version                   Print version and exit
 -w, --web-mode                  Start Links in web mode
```

Resolves #247, resolves #290, resolves #314, resolves #537, resolves #551.

dhil commented:
I should add that I have also garbage collected unused settings. It turned out we had several settings defined in `Basicsettings.ml` that were never referenced. With localised settings, as in this patch, future dead settings will show up as dead code warnings during compilation, if they aren't declared in the interface of the compilation unit.

dhil commented:
Another highlight, that I forgot to mention above, is that settings can have dependencies now. For examples `session_exceptions` depends on `enable_handlers`. Now, setting `session_exceptions=true` will transitively set `enable_handlers` too.

dhil commented:
The GADT code in `Settings` hits a limitation of GADTs prior to OCaml `4.08.x` which has to due with or-patterns. There are two possible solutions: either duplicate the code, or bump the required version of the OCaml compiler to `4.08.1`.

SimonJF commented:
Very nice. Very, very nice. Thank you! Review coming this afternoon.

dhil commented on core/settings.ml:
```diff
@@ -32,193 +142,1059 @@ let parse_bool = function
+    time: int;
+    argument: raw_arg }
+
+(* let string_of_unhandled { time; argument; _ } =
```
Yes.. But it may be useful to keep if you want to debug it in the future :-) I am indifferent as to whether to keep it.

dhil commented on core/settings.ml:
```diff
@@ -1,29 +1,139 @@
+let _ = Printexc.set_uncaught_exception_handler default_handler
+
+let no_conv : string -> 'a
+  = fun _ -> failwith "No converter"
```
I will see whether I can move the global exception handling stuff into `Errors` too....

dhil replied:
OK. So I looked into. It is rather difficult as it introduces a cyclic dependency between errors and settings, because `SourceCode` is included in `Errors`.

dhil replied:
So I will rather keep it as is for now, then maybe we can refactor it in the future. Also note, the `No_converter` error will be a graceful, it will be part of a parse failure message.

SimonJF replied:
OK, that's fine. I have run into this cyclic dependency issue a few times, so it's likely worth doing something about it, but that's for another issue. Thanks for checking.

dhil commented on core/lib.ml:
```diff
@@ -1576,7 +1576,15 @@ let env : (string * (located_primitive * Types.datatype * pure)) list = [
+    (`Server
+       (`PFun (fun _ _ ->
+            Value.(box_list (List.map box_string (Settings.get_rest_arguments ()))))),
+     datatype "() ~> [String]",
```
I think it is best to provide a list of strings, we can easily implement a library that does the proper POSIX parsing on top.

SimonJF replied:
Yeah, and that's for a different patch. Agreed. Could you make an issue though?

dhil commented on core/irtojs.ml:
```diff
@@ -1257,12 +1255,20 @@ end = functor (K : CONTINUATION) -> struct
+let backend =
+  Settings.(option ~default:(Some "cps") "js_compiler"
+            |> privilege `System
+            |> synopsis "Selects the JavaScript compiler"
```
I will add that to `hint`. I will look into enriching the output of `@help` to display the synopsis as well as value/argument hints.

dhil commented on core/basicsettings.ml:
```diff
@@ -1,341 +1,138 @@
-  let port = Settings.add_int ("port", 8080, `User)
-  let external_base_url = Settings.add_string("external_base_url", "", `User)
-  let internal_base_url = Settings.add_string("internal_base_url", "", `User)
+  let hostname =
```
Yes, just need to be a bit careful as it introduces a cyclic dependency between webserver, page, and evalir. But it should be fairly easy to break.. I will have a look.

dhil replied:
OK. It is done.

dhil commented on core/basicsettings.ml:
```diff
@@ -1,341 +1,138 @@
+              |> sync)
+end
+
+module RelationalLenses = struct
```
I will leave it to @rudihorn to do that :-) I don't understand the Lens codebase.

rudihorn replied:
Could probably go in `core/lens.ml` or similar, but would still go in the `core` module. Might do this at some stage but would probably let this be merged in first.

dhil replied:
I moved it into `lens.ml`, in doing so I discovered that neither `classic_lenses` or `debug` are referenced elsewhere in the code base. So I am getting rid of them.

dhil replied:
I see that `classic_lenses` is referenced in `evalir` now.

dhil commented on bin/links.ml:
```diff
@@ -30,31 +62,30 @@ let process_exprarg envs expr =
   if should_start_repl then
     begin
-      print_endline (Settings.get_value BS.welcome_note);
+      Printf.printf "%s%!" (val_of (Settings.get BS.welcome_note));
```
I moved it to `repl.ml`.

dhil commented on core/settings.ml:
```diff
@@ -32,193 +142,1059 @@ let parse_bool = function
+ *   = fun xs ->
+ *   List.iter (fun x -> Printf.printf "%s\n%!" (string_of_unhandled x)) xs *)
+
+let order : unhandled list -> unhandled list
```
Yes, I reckon the list will be a singleton typically. Once you call `ensure_all_synchronised` the list may be big, but if that it is case, then it is likely an error is about to be emitted as we synchronise almost every setting upon definition.

SimonJF replied:
Great, thanks.

dhil commented on core/evalir.ml:
```diff
@@ -12,9 +12,27 @@ let internal_error message =
+    with child processes still running.
+*)
+let concurrent_server =
+  Settings.(flag ~default:true "concurrenct_server"
```
OK. I have removed it.

dhil commented:
> One thing I did notice, though, is that this now only compiles on 4.08.1 and above. This isn't in the patch notes!

I wrote that in a subsequent comment (on here) as I only discovered it after submitting to the CI. As explained, prior to OCaml 4.08.x GADT had restricted pattern matching support. I reckon in this concrete instance, the problem can be fixed by duplicating code paths.

Nevertheless, we should probably consider upgrading to at least 4.07.1, since `cohttp` (one of our core dependencies) no longer builds on `<4.07.x`.

jamescheney commented:
I'm also confused about the comment about `cohttp` requiring 4.07, according to this:

https://github.com/mirage/ocaml-cohttp/blob/master/CHANGES.md

`conduit-mirage` now requies OCaml >= 4.07 but I don't think we need that.  All the other `cohttp` and `conduit` packages still seem to be OK with 4.04-4.07.  The fact that `cohttp` is not yet even being tested on OCaml 4.08 also makes me skeptical that jumping to it is not yet a good idea.

dhil commented:
Sorry I did mean `conduit`, which we do depend on:
```
$ opam list --rec --depends-on conduit -i
# Packages matching: installed & rec-depends-on(conduit)
# Name             # Installed # Synopsis
...
links              0.9         The Links Programming Language
```
I read that changelog a couple of weeks ago. I mustn't have internalised that it was `conduit-mirage` which requires a newer version of the OCaml compiler. My apologies for the confusion.

I am indifferent as to whether we bump bound on the OCaml compiler. I am happy to try the code duplication trick.

dhil commented:
OK. Compatibility with `4.06` should be restored now.

dhil commented:
I have also enriched the output of `@help` directive, for example
```
links> @help db_driver_path;
         name: db_driver_path
         type: multi option
      default: ./_build/install/default/share/links-postgresql,./_build/install/default/share/links-sqlite3
current value: ./_build/install/default/share/links-postgresql,./_build/install/default/share/links-sqlite3
               (value description: <dir[,dir']...>)
     synopsis: Search paths for database drivers
```

dhil commented:
@simonjf the latest commit adds a few tests for the new `getArgs` function. I had to amend the `test-harness` script, because the Python function `str.split` would split `"Hello World!"` into a list of two elements, whilst the correct thing to do (in the context of CLI) is create a singleton list.

I think I have addressed all of your comments now.

dhil commented:
@jamescheney @SimonJF thank you. For completeness I have also successfully tested the examples page both from my git check out, and as a pinned opam installation.

jamescheney commented:
Woo hoo!

Did this also close #560 or is there something to do there?  If it's closed, maybe we can end the "settings" project...

dhil commented:
Oh yes! That is included too.
